tests/cecil-tests/GenericPInvokes.cs

```diff
@@ -0,0 +1,283 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using NUnit.Framework;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+#nullable enable
+
+namespace Cecil.Tests {
+	enum GenericCheckResult {
+		Ok,
+		ContainsGenerics,
+		HasSuspectBranchTarget,
+		HasWeirdInstruction,
+	}
+
+	struct MethodAndResult {
+		public MethodDefinition Method;
+		public GenericCheckResult Result;
+	}
+
+	[TestFixture]
+	public class GenericPInvokesTest {
+		[TestCaseSource (typeof (Helper), nameof (Helper.NetPlatformImplementationAssemblies))]
+		public void CheckSetupBlockUnsafeUsage (string assemblyPath)
+		{
+			// this scans the specified assmebly for all methods
+			// that call SetupBlockUnsafe and then scans the method
+			// to see if the first argument to SetupBlockUnsafe is
+			// a generic delegate type.
+			//
+			// the results are not just a list of booleans.
+			// some of the results are indeterminant.
+			// for example, if the call to SetupBlockUnsafe is
+			// also a branch target within the method then all
+			// bets are off for what's on the stack. If the call
+			// to pass the last argument is also a branch target,
+			// then all bets are off. Why? Imagine this scenario:
+			// SetupBlockUnsafe (someCondition ? val1 : val2, arg2);
+			// This means that the actual argument could be one of
+			// two things and it's probably better for a human to
+			// look at those cases instead of this code.
+			//
+			// most of the code generated by the C# compiler is
+			// pretty boring, which means that there is a very
+			// limited set of instructions that get used to set up
+			// this call. They amount to Ldarg, Ldvar, Ldfld variants
+			// which we like. The only thing that's "funny" is that
+			// when the method is an instance method, then
+			// arg 0 is the "this" pointer which is present on the
+			// stack, but isn't in the formal parameter this.
+			// So there's a little juggling to make sure we don't
+			// look past the array of parameters.
+
+			var assembly = Helper.GetAssembly (assemblyPath, readSymbols: true);
+			var callsToSetupBlock = AllSetupBlocks (assembly);
```

Comment:
Maybe assert that there are more than 0 calls? That way we protect against a typo in the test where it doesn't end up testing anything at all.